### ISSUE TITLE
Docker: Unable to build Javascript starter kit using Docker Instance

### ISSUE THREAD

**Issue by byteshiva** [fastly/cli#515]
Unable to build Javascript starter kit using Docker Instance

**Steps to reproduce:**

Step 1:
get Docker file

```
wget https://raw.githubusercontent.com/fastly/cli/main/Dockerfile-assemblyscript
```
Refer:https://github.com/fastly/cli/blob/main/Dockerfile-assemblyscript

Step 2:
Build Docker 

```
docker build -t fastly/cli/javascript . -f ./Dockerfile-assemblyscript
docker run -v $PWD:/app -it fastly/cli/javascript
```

Step 3:
Initialize Javascript starter kit
```
sample-proj docker run -v $PWD:/app -it --rm fastly/cli/javascript compute init --from=https://github.com/fastly/compute-starter-kit-javascript-default  

Creating a new Compute@Edge project (using --from to locate package template).

Press ^C at any time to quit.

Name: [app] 
Description: Client
Author: Test

✓ Initializing...
✓ Fetching package template...
✓ Updating package manifest...
✓ Initializing package...

Initialized package app to:
	/app

To publish the package (build and deploy), run:
	fastly compute publish

To learn about deploying Compute@Edge projects using third-party orchestration tools, visit:
	https://developer.fastly.com/learning/integrations/orchestration/


SUCCESS: Initialized package app
```

Build javascript project
Step 4:
```
➜  sample-proj docker run -v $PWD:/app -it --rm fastly/cli/javascript compute build                        
✓ Initializing...
✓ Verifying package manifest...
✓ Verifying local javascript toolchain...
✗ Building package using javascript toolchain...

ERROR: error during execution process:
     
     > compute-starter-kit-javascript-default@0.2.0 prebuild
     > webpack
     
     assets by status 0 bytes [cached] 1 asset
     
     WARNING in configuration
     The 'mode' option has not been set, webpack will fallback to 'production' for this
     value.
     Set 'mode' option to 'development' or 'production' to enable defaults for each
     environment.
     You can also set it to 'none' to disable any default behavior. Learn more:
     https://webpack.js.org/configuration/mode/
     
     ERROR in main
     Module not found: Error: Can't resolve './src/index.js' in '/app'
     resolve './src/index.js' in '/app'
       using description file: /app/package.json (relative path: .)
         Field 'browser' doesn't contain a valid alias configuration
         using description file: /app/package.json (relative path: ./src/index.js)
           no extension
             Field 'browser' doesn't contain a valid alias configuration
             /app/src/index.js doesn't exist
           .js
             Field 'browser' doesn't contain a valid alias configuration
             /app/src/index.js.js doesn't exist
           .json
             Field 'browser' doesn't contain a valid alias configuration
             /app/src/index.js.json doesn't exist
           .wasm
             Field 'browser' doesn't contain a valid alias configuration
             /app/src/index.js.wasm doesn't exist
           as directory
             /app/src/index.js doesn't exist
     
     webpack 5.64.0 compiled with 1 error and 1 warning in 202 ms
     

exit status 1.

If you believe this error is the result of a bug, please file an issue: https://github.com/fastly/cli/issues/new?labels=bug&template=bug_report.md
➜  sample-proj 
```

NOTE:

Noticed the docker starter kit project initialization and build steps are writing files in the host as `root`. 

```
➜  sample-proj ll        
total 116K
-rw-r--r--   1 root root  245 Jan 19 11:31 fastly.toml
drwxr-xr-x 100 kapm kapm 4.0K Jan 19 11:31 node_modules
-rw-r--r--   1 root root  91K Jan 19 11:31 npm-shrinkwrap.json
-rw-r--r--   1 root root  848 Jan 19 11:31 package.json
-rw-r--r--   1 root root 1.8K Jan 19 11:31 README.md
drwxr-x---   2 root root 4.0K Jan 19 11:31 src
-rw-r--r--   1 root root  899 Jan 19 11:31 webpack.config.js

```

Refer:
https://github.com/fastly/cli/issues/513

**Comment by byteshiva** [comment]
Refer: https://stackoverflow.com/questions/51949041/module-not-found-error-cant-resolve-app-in-home-user-desktop-my-app-src

To fix build errors in webpack, I added the below configuration in webpack.
```
    module.exports = {
  //...
  resolve: {
    extensions: ['.js', '.jsx']
  }
};
```

```
➜  sample-proj docker run -v $PWD:/app -it --rm fastly/cli/javascript compute build 
✓ Initializing...
✓ Verifying package manifest...
✓ Verifying local javascript toolchain...
✓ Building package using javascript toolchain...
✓ Creating package archive...
```

**Comment by byteshiva** [comment]
Error: Unable to serve JavaScript Project Failed to run using Viceroy!

```
➜  sample-proj docker run -v $PWD:/app -it --rm fastly/cli/javascript compute serve 
✓ Initializing...
✓ Verifying package manifest...
✓ Verifying local javascript toolchain...
✓ Building package using javascript toolchain...
✓ Creating package archive...

SUCCESS: Built package 'app' (pkg/app.tar.gz)

✓ Checking latest Viceroy release...
✓ Fetching latest Viceroy release...
✓ Running local server...

error: Invalid value for '<input>': No such file or directory (os error 2)

ERROR: error during execution process:

error: Invalid value for '<input>': No such file or directory (os error 2).

If you believe this error is the result of a bug, please file an issue: https://github.com/fastly/cli/issues/new?labels=bug&template=bug_report.md
#
```

**Comment by Integralist** [comment]
Hi @byteshiva I've just tried the AssemblyScript Dockerfile but with the node version set to `latest` (which is synonymous with `17`) and it looks to both build and serve fine using the default JavaScript starter kit...

```bash
~/Code/go/testing-cli  Wed 19 Jan 2022 10:10:59 GMT
➜  docker run -v $PWD:/app -it fastly/cli/assemblyscript compute init --language javascript

Creating a new Compute@Edge project.

Press ^C at any time to quit.

Name: [app]
Description:
Author:
Starter kit:
[1] Default starter for JavaScript
    A basic starter kit that demonstrates routing, simple synthetic responses and
    overriding caching rules.
    https://github.com/fastly/compute-starter-kit-javascript-default
Choose option or paste git URL: [1]

✓ Initializing...
✓ Fetching package template...
✓ Updating package manifest...
✓ Initializing package...

Initialized package app to:
        /app

To publish the package (build and deploy), run:
        fastly compute publish

To learn about deploying Compute@Edge projects using third-party orchestration tools, visit:
        https://developer.fastly.com/learning/integrations/orchestration/


SUCCESS: Initialized package app

~/Code/go/testing-cli is 📦 v0.2.0 via  v16.13.0  Wed 19 Jan 2022 10:12:00 GMT
➜  docker run -v $PWD:/app -it fastly/cli/assemblyscript compute serve
✓ Initializing...
✓ Verifying package manifest...
✓ Verifying local javascript toolchain...
✓ Building package using javascript toolchain...
✓ Creating package archive...

SUCCESS: Built package 'app' (pkg/app.tar.gz)

✓ Checking latest Viceroy release...
✓ Fetching latest Viceroy release...
✓ Running local server...

Jan 19 10:12:36.254  WARN no backend definitions found in /app/fastly.toml
Jan 19 10:12:36.254  INFO Listening on http://127.0.0.1:7676
```

**Comment by byteshiva** [comment]
No, I'm still getting errors,

```
➜  docker more Dockerfile 
FROM node:latest
LABEL maintainer="Fastly OSS <oss@fastly.com>"

RUN apt-get update && apt-get install -y curl jq
RUN export FASTLY_CLI_VERSION=$(curl --silent https://api.github.com/repos/fastly/cli/releases/latest | jq -r .tag_name | cut -d 'v' -f 2) \
  GOARCH=$(dpkg --print-architecture) \
  && curl -vL "https://github.com/fastly/cli/releases/download/v${FASTLY_CLI_VERSION}/fastly_v${FASTLY_CLI_VERSION}_linux-$GOARCH.tar.gz" -o fastly.tar.gz \
  && curl -vL "https://github.com/fastly/cli/releases/download/v${FASTLY_CLI_VERSION}/fastly_v${FASTLY_CLI_VERSION}_SHA256SUMS" -o sha256sums \
  && dlsha=$(shasum -a 256 fastly.tar.gz | cut -d " " -f 1) && expected=$(cat sha256sums | awk -v pat="$dlsha" '$0~pat' | cut -d " " -f 1) \
  && if [[ "$dlsha" != "$expected" ]]; then echo "shasums don't match" && exit 1; fi

RUN tar -xzf fastly.tar.gz --directory /usr/bin && rm fastly.tar.gz

WORKDIR /app
ENTRYPOINT ["/usr/bin/fastly"]
CMD ["--help"]

# docker build -t fastly/cli/assemblyscript . -f ./Dockerfile-assemblyscript
# docker run -v $PWD:/app -it fastly/cli/assemblyscript
➜  docker cd - 
total 0
➜  sample-js docker run -v $PWD:/app -it --rm fastly/cli/javascript com           
➜  sample-js docker run -v $PWD:/app -it --rm fastly/cli/javascript compute init --language javascript


Creating a new Compute@Edge project.

Press ^C at any time to quit.

Name: [app] 
Description: 
Author: 
Starter kit:
[1] Default starter for JavaScript
    A basic starter kit that demonstrates routing, simple synthetic responses and
    overriding caching rules.
    https://github.com/fastly/compute-starter-kit-javascript-default
Choose option or paste git URL: [1] 

✓ Initializing...
✓ Fetching package template...
✓ Updating package manifest...
✓ Initializing package...

Initialized package app to:
	/app

To publish the package (build and deploy), run:
	fastly compute publish

To learn about deploying Compute@Edge projects using third-party orchestration tools, visit:
	https://developer.fastly.com/learning/integrations/orchestration/


SUCCESS: Initialized package app
➜  sample-js docker run -v $PWD:/app -it --rm fastly/cli/javascript compute serve                 
✓ Initializing...
✓ Verifying package manifest...
✓ Verifying local javascript toolchain...
✗ Building package using javascript toolchain...

ERROR: error during execution process:
     
     > compute-starter-kit-javascript-default@0.2.0 prebuild
     > webpack
     
     assets by status 0 bytes [cached] 1 asset
     
     WARNING in configuration
     The 'mode' option has not been set, webpack will fallback to 'production' for this
     value.
     Set 'mode' option to 'development' or 'production' to enable defaults for each
     environment.
     You can also set it to 'none' to disable any default behavior. Learn more:
     https://webpack.js.org/configuration/mode/
     
     ERROR in main
     Module not found: Error: Can't resolve './src/index.js' in '/app'
     resolve './src/index.js' in '/app'
       using description file: /app/package.json (relative path: .)
         Field 'browser' doesn't contain a valid alias configuration
         using description file: /app/package.json (relative path: ./src/index.js)
           no extension
             Field 'browser' doesn't contain a valid alias configuration
             /app/src/index.js doesn't exist
           .js
             Field 'browser' doesn't contain a valid alias configuration
             /app/src/index.js.js doesn't exist
           .json
             Field 'browser' doesn't contain a valid alias configuration
             /app/src/index.js.json doesn't exist
           .wasm
             Field 'browser' doesn't contain a valid alias configuration
             /app/src/index.js.wasm doesn't exist
           as directory
             /app/src/index.js doesn't exist
     
     webpack 5.64.0 compiled with 1 error and 1 warning in 241 ms
     

exit status 1.

If you believe this error is the result of a bug, please file an issue: https://github.com/fastly/cli/issues/new?labels=bug&template=bug_report.md
➜  sample-js 

```

**Comment by byteshiva** [comment]
Note:
The docker run has changed host filesystem ownership to `root`. 

```
➜  sample-js ls -rlt
total 116
-rw-r--r--   1 root root  1754 Jan 19 16:36 README.md
-rw-r--r--   1 root root   848 Jan 19 16:36 package.json
drwxr-x---   2 root root  4096 Jan 19 16:36 src
-rw-r--r--   1 root root   899 Jan 19 16:36 webpack.config.js
-rw-r--r--   1 root root   338 Jan 19 16:36 fastly.toml
-rw-r--r--   1 root root 92661 Jan 19 16:37 npm-shrinkwrap.json
drwxr-xr-x 100 kptm kptm  4096 Jan 19 16:37 node_modules
➜  sample-js 
````

**Comment by Integralist** [comment]
FYI I've just merged this PR https://github.com/fastly/cli/pull/517 which renames `Dockerfile-assemblyscript` to `Dockerfile-node` (just in case you try pulling the old filename down).

**Comment by Integralist** [comment]
The error you're seeing is interesting because (thanks to using a container) I should in theory see the exact same error as you, but I don't 🤔 

I know you added a `resolve: {extensions: ['.js', '.jsx']}` which at the time helped you get the project to build. Now if I add that to the `webpack.config.js` file provided by the default JS starter kit I still don't get the error you're seeing 🤔

**Comment by byteshiva** [comment]
Still getting the same error for Dockerfile-node as well:

Error details available 
https://gist.github.com/byteshiva/2517c5f38813f19dc1c0e8fda4dceb19

ps:
I've installed node on my local machine, I think that should not affect the docker instance!

**Comment by byteshiva** [comment]
I ran it on Google Cloud Shell, 

Threw me different errors:

Details available at
https://gist.github.com/byteshiva/057cec3a644f2db101298c5d07d85e6f

**Comment by byteshiva** [comment]
I'm able to replicate it on Google Cloud Shell.

![image](https://user-images.githubusercontent.com/10637299/150125832-5ed63563-b74d-4444-8c71-93d34d2c982a.png)

**Comment by Integralist** [comment]
Hmm, interesting. I had an internal comment to say...

> They probably have cached layers. Would suggest using `--no-cache` flag on the `docker build` command.

But seeing your latest comment re: Google Cloud Shell that seems unlikely to help as that would already be a fresh set of layers.

**Comment by byteshiva** [comment]
> Hmm, interesting. I had an internal comment to say...
> 
> > They probably have cached layers. Would suggest using `--no-cache` flag on the `docker build` command.

It didn't help!

Details available at 
https://gist.github.com/byteshiva/b3c4b2ecc4bb22e5f09fb8619bccf06a

**Comment by Integralist** [comment]
re: this JS build issue -- can you confirm the code builds successfully _outside_ of Docker using a non-dockerised version of the CLI. Thanks.

**Comment by byteshiva** [comment]
I ran it on google cloudshell, Now it seems viceroy depends on "GLIBC_2.29".

```
kptm@cloudshell:~/sample-js/sample-1 (quickstart-54645622)$ fastly compute serve -v
Fastly API token not provided
Fastly API endpoint: https://api.fastly.com
Verifying package manifest...
Verifying local javascript toolchain...
Checking if npm is installed...
Found npm at /usr/local/nvm/versions/node/v12.14.1/bin/npm
Found package.json at /home/kptm/sample-js/sample-1/package.json
Checking if @fastly/js-compute is installed...
Found js-compute-runtime at /home/kptm/sample-js/sample-1/node_modules/.bin/js-compute-runtime
Building package using javascript toolchain...

> compute-starter-kit-javascript-default@0.2.0 prebuild
> webpack

asset index.js 37.7 KiB [compared for emit] [minimized] (name: main) 1 related asset
runtime modules 495 bytes 2 modules
modules by path ./node_modules/core-js/internals/*.js 65.3 KiB 95 modules
modules by path ./node_modules/core-js/modules/*.js 48.6 KiB
  ./node_modules/core-js/modules/web.url.js 32.5 KiB [built] [code generated]
  ./node_modules/core-js/modules/web.url-search-params.js 12.6 KiB [built] [code generated]
  ./node_modules/core-js/modules/web.url.to-json.js 324 bytes [built] [code generated]
  ./node_modules/core-js/modules/es.string.iterator.js 1.06 KiB [built] [code generated]
  ./node_modules/core-js/modules/es.array.iterator.js 2.1 KiB [built] [code generated]
modules by path ./src/ 2.8 KiB
  ./src/index.js 2.14 KiB [built] [code generated]
  ./src/welcome-to-compute@edge.html 678 bytes [built] [code generated]
./node_modules/core-js/web/url.js 184 bytes [built] [code generated]

WARNING in configuration
The 'mode' option has not been set, webpack will fallback to 'production' for this value.
Set 'mode' option to 'development' or 'production' to enable defaults for each environment.
You can also set it to 'none' to disable any default behavior. Learn more: https://webpack.js.org/configuration/mode/

webpack 5.64.0 compiled with 1 warning in 2482 ms

> compute-starter-kit-javascript-default@0.2.0 build
> js-compute-runtime --skip-pkg bin/index.js bin/main.wasm


Creating package archive...

SUCCESS: Built package 'sample-1' (pkg/sample-1.tar.gz)

Fetching latest Viceroy release...
Running local server...
Wasm file: bin/main.wasm
Manifest: /home/kptm/sample-js/sample-1/fastly.toml

/home/kptm/.config/fastly/viceroy: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by /home/kptm/.config/fastly/viceroy)

ERROR: error during execution process:

/home/kptm/.config/fastly/viceroy: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by /home/kptm/.config/fastly/viceroy).

If you believe this error is the result of a bug, please file an issue: https://github.com/fastly/cli/issues/new?labels=bug&template=bug_report.md
```

**Comment by byteshiva** [comment]
I tweaked the Dockerfile to make it work. I commented on both ENTRYPOINT and CMD in Dockerfile.

Now instead of running `docker run -v $PWD:/app -it -p 7676:7676 fastly/cli/node compute serve --addr="0.0.0.0:7676"`

I did  
`docker build -t fastly/cli/node_cli . -f ./Dockerfile-node`
`docker run  -it fastly/cli/node_cli /bin/bash`

I logged into the container and executed the vanilla javascript project. And it works fine. No error thrown for Viceroy


Dockerfile.node
```
FROM node:latest
LABEL maintainer="Fastly OSS <oss@fastly.com>"

RUN apt-get update && apt-get install -y curl jq
RUN export FASTLY_CLI_VERSION=$(curl --silent https://api.github.com/repos/fastly/cli/releases/latest | jq -r .tag_name | cut -d 'v' -f 2) \
  GOARCH=$(dpkg --print-architecture) \
  && curl -vL "https://github.com/fastly/cli/releases/download/v${FASTLY_CLI_VERSION}/fastly_v${FASTLY_CLI_VERSION}_linux-$GOARCH.tar.gz" -o fastly.tar.gz \
  && curl -vL "https://github.com/fastly/cli/releases/download/v${FASTLY_CLI_VERSION}/fastly_v${FASTLY_CLI_VERSION}_SHA256SUMS" -o sha256sums \
  && dlsha=$(shasum -a 256 fastly.tar.gz | cut -d " " -f 1) && expected=$(cat sha256sums | awk -v pat="$dlsha" '$0~pat' | cut -d " " -f 1) \
  && if [[ "$dlsha" != "$expected" ]]; then echo "shasums don't match" && exit 1; fi

RUN tar -xzf fastly.tar.gz --directory /usr/bin && rm fastly.tar.gz

WORKDIR /app
# ENTRYPOINT ["/usr/bin/fastly"]
# CMD ["--help"]
```

docker build -t fastly/cli/node_cli . -f ./Dockerfile-node


```
root@c32a8816b689:/app# fastly compute init --language javascript

Creating a new Compute@Edge project.

Press ^C at any time to quit.

Name: [app] 
Description: 
Author: 
Starter kit:
[1] Default starter for JavaScript
    A basic starter kit that demonstrates routing, simple synthetic responses and
    overriding caching rules.
    https://github.com/fastly/compute-starter-kit-javascript-default
Choose option or paste git URL: [1] 

✓ Initializing...
- Fetching package template...
✓ Fetching package template...
✓ Updating package manifest...
✓ Initializing package...

Initialized package app to:
        /app

To publish the package (build and deploy), run:
        fastly compute publish

To learn about deploying Compute@Edge projects using third-party orchestration tools, visit:
        https://developer.fastly.com/learning/integrations/orchestration/


SUCCESS: Initialized package app
root@c32a8816b689:/app#
root@c32a8816b689:/app# fastly compute serve
✓ Initializing...
✓ Verifying package manifest...
✓ Verifying local javascript toolchain...
✓ Building package using javascript toolchain...
✓ Creating package archive...

SUCCESS: Built package 'app' (pkg/app.tar.gz)

✓ Checking latest Viceroy release...
✓ Fetching latest Viceroy release...
✓ Running local server...

Jan 20 13:55:28.301  WARN no backend definitions found in /app/fastly.toml
Jan 20 13:55:28.302  INFO Listening on http://127.0.0.1:7676
```

**Comment by byteshiva** [comment]
I think I got it working for Dockerfile. 

This issue is related to https://blog.gougousis.net/file-permissions-the-painful-side-of-docker/

You'd need to pass `--user $(id -u)` in all your Dockerfiles.

```
➜  sample1 docker run --user $(id -u) -v $PWD:/app -it --rm fastly/cli/assemblyscript:17 compute init --language javascript 


Creating a new Compute@Edge project.

Press ^C at any time to quit.

Name: [app] 
Description: 
Author: 
Starter kit:
[1] Default starter for JavaScript
    A basic starter kit that demonstrates routing, simple synthetic responses and
    overriding caching rules.
    https://github.com/fastly/compute-starter-kit-javascript-default
Choose option or paste git URL: [1] 

✓ Initializing...
- Fetching package template...
✓ Fetching package template...
✓ Updating package manifest...
- npm WARN EBADENGINE }
✓ Initializing package...

Initialized package app to:
	/app

To publish the package (build and deploy), run:
	fastly compute publish

To learn about deploying Compute@Edge projects using third-party orchestration tools, visit:
	https://developer.fastly.com/learning/integrations/orchestration/


SUCCESS: Initialized package app
➜  sample1 docker run --user $(id -u) -v $PWD:/app -it -p 7676:7676 --rm fastly/cli/assemblyscript:17 compute serve --addr="0.0.0.0:7676"
✓ Initializing...
✓ Verifying package manifest...
✓ Verifying local javascript toolchain...
✓ Building package using javascript toolchain...
✓ Creating package archive...

SUCCESS: Built package 'app' (pkg/app.tar.gz)

✓ Checking latest Viceroy release...
✓ Fetching latest Viceroy release...
✓ Running local server...

Jan 20 14:31:01.093  WARN no backend definitions found in /app/fastly.toml
Jan 20 14:31:01.093  INFO Listening on http://0.0.0.0:7676

```

**Comment by Integralist** [comment]
I'm glad that's working for you. I find that I get an error if I try to run the container with `--user $(id -u)`...

```
ERROR: mkdir /.config: permission denied.
```

**Comment by byteshiva** [comment]
> Thanks @byteshiva I'm glad you managed to get it working for you.
> 
> I actually find it errors if I try to run the container with `--user $(id -u)`.
> 
> ```
> ERROR: mkdir /.config: permission denied.
> ```

Did you create a new folder and run fastly `init and build and serve` for language javascript?

**Comment by Integralist** [comment]
So I had a pre-existing folder which I ran...

```bash
$ rm -rf *
$ docker run -v $PWD:/app -it -p 7676:7676 --user $(id -u) fastly/cli/node compute init --language javascript
```

The CLI then attempts to create a config directory which in this case fails. But if I omit the `--user` flag then it'll succeed and initialise a new project no problem.

**Comment by byteshiva** [comment]
> 

rm -rf will not delete hidden files. You would need to create a new folder altogether and try it again

```
➜  js2 mkdir .config 
➜  js2 rm -rf * 
zsh: sure you want to delete all the files in /home/siva/developments/wasm-workspaces/fastly-workspaces/fastly-docker/javascript/js2 [yn]? %                                                                                     ➜  js2 rm -rf * 
zsh: sure you want to delete all the files in /home/siva/developments/wasm-workspaces/fastly-workspaces/fastly-docker/javascript/js2 [yn]? y
zsh: no matches found: *
➜  js2 
➜  js2 ll
total 0
➜  js2 ls -rlt 
total 0
➜  js2 cd .
./        ../       .config/



```

**Comment by Integralist** [comment]
Just to quickly highlight that the use of `--user` didn't work for me, even in a fresh directory (but remove `--user` and it worked fine)...

<img width="1178" alt="Screenshot 2022-01-20 at 15 57 46" src="https://user-images.githubusercontent.com/180050/150374707-e12f5e86-a4a0-43dd-b880-3e50b3b2fd49.png">

**Comment by Integralist** [comment]
Also potentially of interest to you, is https://github.com/fastly/cli/pull/521 which was suggested internally we might want to add.

**Comment by byteshiva** [comment]
> Just to quickly highlight that the use of `--user` didn't work for me, even in a fresh directory (but remove `--user` and it worked fine)...
> 
> <img width="1178" alt="Screenshot 2022-01-20 at 15 57 46" src="https://user-images.githubusercontent.com/180050/150374707-e12f5e86-a4a0-43dd-b880-3e50b3b2fd49.png">
> 

Could you let me know the file permission/ownership set on HOST filesystem, it is root:root

**Comment by Integralist** [comment]
```
During fastly init, I see all the files created on HOST are set as root, that makes it read-only.
```

I imagine then that #521 should help as it's switching to a non-root user.

**Comment by byteshiva** [comment]
> > During fastly init, I see all the files created on HOST are set as root, that makes it read-only.
> 
> I imagine then that #521 should help as it's switching to a non-root user.

This PR fails to work in my case

```
 docker run  -v $PWD:/app -it --rm fastly/cli/node_cls compute init --language javascript    

Creating a new Compute@Edge project.

Press ^C at any time to quit.


ERROR: error creating file in package destination: open /app/tmp_c309f53cd4e29058f7768295c5efd30f: permission denied.

If you believe this error is the result of a bug, please file an issue: https://github.com/fastly/cli/issues/new?labels=bug&template=bug_report.md
➜  js3 

```

**Comment by byteshiva** [comment]
I tried it on google cloud shell as well, It threw me the following errors.

```
* Connection #1 to host objects.githubusercontent.com left intact
/bin/sh: 1: [[: not found
Removing intermediate container a643276bffb0
 ---> 7aff9e9eaaac
Step 5/10 : RUN tar -xzf fastly.tar.gz --directory /usr/bin && rm fastly.tar.gz
 ---> Running in efffecfa9ea5
Removing intermediate container efffecfa9ea5
 ---> 82785fde7afc
Step 6/10 : RUN adduser fastly
 ---> Running in b47747567bef
Adding user `fastly' ...
Adding new group `fastly' (1001) ...
Adding new user `fastly' (1001) with group `fastly' ...
Creating home directory `/home/fastly' ...
Copying files from `/etc/skel' ...
New password: Password change has been aborted.
passwd: Authentication token manipulation error
passwd: password unchanged
Use of uninitialized value $answer in chop at /usr/sbin/adduser line 556.
Use of uninitialized value $answer in pattern match (m//) at /usr/sbin/adduser line 557.
Try again? [y/N] Changing the user information for fastly
Enter the new value, or press ENTER for the default
        Full Name []:   Room Number []:         Work Phone []:  Home Phone []:  Other []: Use of uninitialized value $answer in chop at /usr/sbin/adduser line 582.
Use of uninitialized value $answer in pattern match (m//) at /usr/sbin/adduser line 583.
Is the information correct? [Y/n] Removing intermediate container b47747567bef
 ---> 732599d901d3
Step 7/10 : USER fastly
 ---> Running in 03fdaa7549ff
Removing intermediate container 03fdaa7549ff
 ---> ff89f7a02357
Step 8/10 : WORKDIR /app
 ---> Running in 4c77a13112fd
Removing intermediate container 4c77a13112fd
 ---> 1cc50ca0085b
Step 9/10 : ENTRYPOINT ["/usr/bin/fastly"]
 ---> Running in 0f56a11ad198
Removing intermediate container 0f56a11ad198
 ---> 8ee725259771
Step 10/10 : CMD ["--help"]
 ---> Running in b42dcac8938c
Removing intermediate container b42dcac8938c
 ---> b9acbb3fb02f
Successfully built b9acbb3fb02f
Successfully tagged fastly/cli/node_cls:latest

```

```
kptm@cloudshell:~/temp1 (eeter-noder-21234)$ mkdir jscloud1
kptm@cloudshell:~/temp1 (eeter-noder-21234)$ cd jscloud1/
kptm@cloudshell:~/temp1/jscloud1 (eeter-noder-21234)$ docker run  -v $PWD:/app -it --rm fastly/cli/node_cls compute init --language javascript



Creating a new Compute@Edge project.

Press ^C at any time to quit.


ERROR: error creating file in package destination: open /app/tmp_50ca9ce53ecdc295f18a3e7fc5241e28: permission denied.

If you believe this error is the result of a bug, please file an issue: https://github.com/fastly/cli/issues/new?labels=bug&template=bug_report.md
```

```
kptm@cloudshell:~/temp1/jscloud1 (eeter-noder-21234)$ docker run  -v $PWD:/app -it --rm fastly/cli/node_cls compute init --language javascript -v
Fastly API token not provided
Fastly API endpoint: https://api.fastly.com

Creating a new Compute@Edge project.

Press ^C at any time to quit.

--directory not specified, using current directory

ERROR: error creating file in package destination: open /app/tmp_3d4f5fd2dcd1ce05ed1dbdbaf4178477: permission denied.

If you believe this error is the result of a bug, please file an issue: https://github.com/fastly/cli/issues/new?labels=bug&template=bug_report.md
kptm@cloudshell:~/temp1/jscloud1 (eeter-noder-21234)$
```

**Comment by byteshiva** [comment]
> I think I got it working for Dockerfile.
> 
> This issue is related to https://blog.gougousis.net/file-permissions-the-painful-side-of-docker/
> 
> You'd need to pass `--user $(id -u)` when running docker instance using `docker run `
> 
> ```
> ➜  sample1 docker run --user $(id -u) -v $PWD:/app -it --rm fastly/cli/assemblyscript:17 compute init --language javascript 
> 
> 
> Creating a new Compute@Edge project.

It works on Google cloud shell if I provide `--user $(id -u):$(id -g) ` and built image by removing `RUN adduser fastly and USER fastly` from Dockerfile.

```
kptm@cloudshell:~/temp1/jscloud2 (eteran-trn-21144)$ docker run --user $(id -u):$(id -g)  -v $PWD:/app -it --rm fastly/cli/node compute init --language javascript -v
Fastly API token not provided
Fastly API endpoint: https://api.fastly.com

Creating a new Compute@Edge project.

Press ^C at any time to quit.

--directory not specified, using current directory
Name: [app] 
Description: 
Author: 
Starter kit:
[1] Default starter for JavaScript
    A basic starter kit that demonstrates routing, simple synthetic responses and
    overriding caching rules.
    https://github.com/fastly/compute-starter-kit-javascript-default
Choose option or paste git URL: [1] 

Fetching package template...
Updating package manifest...
Setting package name in manifest to "app"...
Setting authors in manifest to ...
Setting language in manifest to javascript...
Initializing package...
Checking if npm is installed...
Found npm at /usr/local/bin/npm
Found package.json at /app/package.json
Installing package dependencies...
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'compute-starter-kit-javascript-default@0.2.0',
npm WARN EBADENGINE   required: { node: '^16' },
npm WARN EBADENGINE   current: { node: 'v17.4.0', npm: '8.3.1' }
npm WARN EBADENGINE }

added 123 packages, and audited 124 packages in 5s

18 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
npm notice
npm notice New patch version of npm available! 8.3.1 -> 8.3.2
npm notice Changelog: <https://github.com/npm/cli/releases/tag/v8.3.2>
npm notice Run `npm install -g npm@8.3.2` to update!
npm notice

Initialized package app to:
        /app

To publish the package (build and deploy), run:
        fastly compute publish

To learn about deploying Compute@Edge projects using third-party orchestration tools, visit:
        https://developer.fastly.com/learning/integrations/orchestration/


SUCCESS: Initialized package app

```
```
kptm@cloudshell:~/temp1/jscloud2 (eteran-trn-21144)$ docker run --user $(id -u):$(id -g) -p 7676:7676  -v $PWD:/app -it --rm fastly/cli/node compute serve --addr="0.0.0.0:7676"
✓ Initializing...
✓ Verifying package manifest...
✓ Verifying local javascript toolchain...
✓ Building package using javascript toolchain...
✓ Creating package archive...

SUCCESS: Built package 'app' (pkg/app.tar.gz)

✓ Checking latest Viceroy release...
✓ Fetching latest Viceroy release...
✓ Running local server...

Jan 21 01:44:57.264  WARN no backend definitions found in /app/fastly.toml
Jan 21 01:44:57.264  INFO Listening on http://0.0.0.0:7676

```

**Comment by Integralist** [comment]
In this case, as the Dockerfiles aren't officially supported and because you have found a solution to your issues. I'm going to close off this issue and #516.
